### PR TITLE
Setting Up a Python Dev Environment with DevContainers and Daytona

### LINKED ISSUE #10: Setting Up a Python Dev Environment with DevContainers and Daytona
### Content Type

Guide

### Article Description

Daytona as endgame Python dev env. how to skip all env managers like venv, etc. and just use Daytona.

Explore the benefits of using DevContainers for Python development and how to set them up step by step. You can use a provided reference repo, or create a new one with some specific spin, etc.

### Target Audience

Python devs, data scientists, ml/ai devs.

### References/Resources

https://github.com/pamelafox/python-project-template
https://www.daytona.io/dotfiles/the-ultimate-guide-to-managing-python-environments

### Examples

_No response_

### Special Instructions

_No response_

### PR DESCRIPTION
**Pull request by Kiran1689**
fix #10 
/claim #10 
### Writing Structure
- [x] Use short sentences and paragraphs, and include bucket brigades.
- [x] Include more than two descriptive H2 headings to organize content.
- [x] Capitalize headings according to the AP Stylebook ([use this tool](https://capitalizemytitle.com/style/AP/))
- [x] Include an introduction with at least two paragraphs before the first H2 section.
- [x] Use appropriate formatting (bold, italic, underline), notes, quotes, TLDRs, and key points.
- [x] Incorporate `code elements` and `Markdown format` where appropriate.
- [x] Ensure at least one visual element per “screen” of the article (screenshots, diagrams, tables, graphs, lists, sidenotes, blockquotes).

### Fact-Checking
- [x] Verify all facts and data points included in the article.

### Asset Management
- [x] Save images in the `/assets` folder.
- [x] Follow naming conventions: `YYYYMMDD_title_of_the_article_IMG_NAME_NO.png`.
- [ ] (Optional) Create a GitHub repo for the code referenced in the article and share it.
- [ ] (Optional) Include a link to this Loom video in the PR comments.

### Interlinking
- [x] Use `CTRL+F` to search for relevant keywords in:
    - [Daytona Dotfiles Sitemap](https://www.daytona.io/sitemap-dotfiles.xml)
    - [Daytona Definitions Sitemap](https://www.daytona.io/sitemap-definitions.xml)
- [x] Interlink keywords using the appropriate URLs.

### Glossary/Definitions
- [ ] Create new definition in `/defitnitions` folder.

### Review and Edit
- [x] Ensure articles address the needs of the target audience and their search intent.
- [x] Read the article out loud to catch any awkward phrasing.
- [x] Run the draft through [Grammarly](https://grammarly.com) or a similar grammar tool.
- [x] Double-check adherence to the style guide and repository guidelines.
- [x] Use the name of the article for the title of the PR.

### CODE REVIEW

**Comment by Kiran1689** [comment]
@mojafa Can you please review?

**Comment by mojafa** [comment]
Slight spacing issue here:

**Comment by mojafa** [comment]
@Kiran1689 your guide works. 

fix the requests then we can merge.

**Comment by Kiran1689** [comment]
> Slight spacing issue here:

@mojafa  The image is not showing for the issue can you please reshare

**Comment by Kiran1689** [comment]
updated @mojafa

**Comment by Kiran1689** [comment]
Anything else @mojafa ??

**Comment by mojafa** [comment]
@Kiran1689 sorry I think my previous comments hadn't gone through.
I noticed this PR has content from your other issue attempts. Please clean up the project, remove any other issue attemps and only have this guide in your  guides section as well as all the assets. 

see below.

Screenshot 2024-11-21 at 22 46 50

**Comment by mojafa** [comment]
only hjave this a=guide and the template as well as the assets, including author section as well

**Comment by Kiran1689** [comment]
@mojafa Sorry, I didn't get that.

What I need to remove? Can you please tell

My branch is up to date with the main branch.. As the main repo contains the my author details I didn't added to this PR again

**Comment by Kiran1689** [comment]
@nkkko @mojafa Can you please look into this

**Comment by Kiran1689** [comment]
@mojafa You can check the files added for this PR here - https://github.com/daytonaio/content/pull/120/files

This PR contains only files related to the issue, no other contribution files.

image

**Comment by Kiran1689** [comment]
@mojafa Can you check this

**Comment by mojafa** [comment]
@Kiran1689  on it

**Comment by mojafa** [comment]
@nkkko PR meets all requirements

**Comment by Kiran1689** [comment]
@nkkko Can you please look into this?

File: guides/20241121_guide_setting_up_a_python_dev_environment_with_devcontainers_and_daytona.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+Daytona will now start creating a workspace by pulling your project and install all required dependencies, once done it will open your project in your default IDE.
+
+![Daytona workspace](assets/setting_up_a_python_dev_environment_with_devcontainers_and_daytona_img8.png)
+
+![Daytona workspace](assets/setting_up_a_python_dev_environment_with_devcontainers_and_daytona_img9.png)
+
+Now, your python dev environmentis ready to use.
+
+### Step 3: Confirmation
+Now, In your VS code run the test cases, you will see all the testcases passed. This means python dev environment has successfull set for your project.
```
Comment (nkkko) [inline guides/20241121_guide_setting_up_a_python_dev_environment_with_devcontainers_and_daytona.md:145]: Code!

File: guides/20241121_guide_setting_up_a_python_dev_environment_with_devcontainers_and_daytona.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+By following this guide, you've set up a fully containerized Python development environment using Daytona and DevContainers. This setup ensures that your projects are reproducible, isolated, and consistent across machines, enabling smooth collaboration and effortless debugging.
+
+## References
+
+- [Python Project Template by Pamela Fox](https://github.com/pamelafox/python-project-template)
+
+- [Daytona Documentation](https://www.daytona.io/dotfiles/the-ultimate-guide-to-managing-python-environments)
+
+
+<!-- Note on Definitions -->
```
Comment (nkkko) [inline guides/20241121_guide_setting_up_a_python_dev_environment_with_devcontainers_and_daytona.md:159]: @mojafa please take more attention, notes should be removed

File: guides/20241121_guide_setting_up_a_python_dev_environment_with_devcontainers_and_daytona.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+Now, your python dev environmentis ready to use.
+
+### Step 3: Confirmation
+Now, In your VS code run the test cases, you will see all the testcases passed. This means python dev environment has successfull set for your project.
+
+![Daytona workspace](assets/setting_up_a_python_dev_environment_with_devcontainers_and_daytona_img10.png)
+
+## Conclusion
+By following this guide, you've set up a fully containerized Python development environment using Daytona and DevContainers. This setup ensures that your projects are reproducible, isolated, and consistent across machines, enabling smooth collaboration and effortless debugging.
```
Comment (nkkko) [inline guides/20241121_guide_setting_up_a_python_dev_environment_with_devcontainers_and_daytona.md:150]: Check the right way of writing Dev Container not DevContainer

**Comment by Kiran1689** [comment]
@mojafa @nkkko I have updated the content and made changes accordingly..

**Comment by Kiran1689** [comment]
@nkkko 👀

**Comment by mojafa** [comment]
Will rereview

**Comment by mojafa** [comment]
@Kiran1689 quite a lot of `DevContainer` all over and you still have notes in your file.

**Comment by Kiran1689** [comment]
@mojafa 
- I have removed the notes and replaced DevContainer with Dev Container all over the article except in image name and file name
- fixed the lint issues in the guide as you mentioned but still its failing bcs of issues in all other files

**Comment by mojafa** [comment]
Screenshot 2024-12-02 at 19 08 43
@Kiran1689 as i figure out where your PR falls short on the Content Compliance Check, please check your guide for spelling and cohesion, use grammarly or hemingways, Like that should be `has successfully` or `was successful`

**Comment by mojafa** [comment]
@Kiran1689 also please run:
`npm install -g markdownlint-cli `and then **npx markdownlint "**/*.md"** to see where your Content Compliance Checks are failing.

but mostly here:
**Heading Structure Errors**

- MD025/single-title/single-h1: Multiple top-level (H1) headings in the same document.
- MD041/first-line-heading/first-line-h1: The first line in a file should be a top-level heading (H1).
- MD022/blanks-around-headings: Headings should be surrounded by blank lines.
- MD036/no-emphasis-as-heading: Emphasis (bold or italic) used instead of a proper heading.
- MD023/heading-start-left: Headings must start at the beginning of the line.

**Formatting and Style Errors**

- MD009/no-trailing-spaces: Trailing spaces at the end of lines.
- MD012/no-multiple-blanks: Multiple consecutive blank lines; typically, only one blank line is expected.
- MD004/ul-style: Unordered list style inconsistency (e.g., using asterisks * instead of dashes -).
- MD007/ul-indent: Incorrect indentation for unordered lists.
- MD010/no-hard-tabs: Hard tabs (\t) used instead of spaces.
- MD034/no-bare-urls: Bare URLs used without proper formatting.
- MD038/no-space-in-code: Spaces inside inline code spans.
- MD029/ol-prefix: Incorrect ordered list item prefix.
- MD039/no-space-in-links: Spaces inside link text.
- MD049/emphasis-style: Inconsistent emphasis style (e.g., mixing underscores _ and asterisks *).

**Comment by Kiran1689** [comment]
@mojafa, I've updated the content to correct grammar and spelling mistakes and resolved all lint issues in my guide. However, it’s still failing due to lint issues in other files, including node modules. Can't fix that.

I noticed that all other PRs are failing the Content Compliance Check.

**Comment by Kiran1689** [comment]
@mojafa Now it's passing all checks, waiting for your approval..

**Comment by mojafa** [comment]
@Kiran1689 here's my feedback:

**Terminology Corrections:**

"DevContainer" vs. "Dev Container": The correct term is "Dev Container." Please update all instances of "DevContainer" to "Dev Container" throughout the article to maintain accuracy.

**Spacing Around Links**: 
There are several instances where spacing is missing around links, such as in the prerequisites section:
Incorrect: `Basic understanding of[Python](definitions/20240820_defintion_python.md)`
Correct: `Basic understanding of [Python](definitions/20240820_defintion_python.md)`

**Lack of Detailed Instructions**: 
The article introduces setting up a Python development environment using Dev Containers and Daytona but lacks in-depth, technical guidance.

Example: Sections like "## Step 1: Preparations" and "### Fork the Template Repository" are brief and do not provide enough detailed steps for a reader to follow effectively.


**Recommendation**: Consider following a structure similar to https://www.daytona.io/dotfiles/the-ultimate-guide-to-managing-python-environments . This guide offers detailed, step-by-step instructions that would greatly benefit readers.


Incomplete Steps: In "### Step 3: Confirmation," the instructions are vague:
"Now, In your VS code run the test cases, you will see all the testcases passed. This means python dev environment has successfull set for your project."
Action Item: Provide explicit instructions on how to run the test cases in VS Code. Include screenshots, commands, or code snippets as necessary to guide the reader.


Rework Section Headings: The current headings like "## Step 1: Preparations" can be made more descriptive. For example:

"## Step 1: Setting Up Your Development Environment"
"### Cloning the Template Repository"
Flow of Content: Ensure that each section logically leads to the next, providing a smooth journey for the reader from start to finish.

Screenshot 2024-12-05 at 01 00 32
Screenshot 2024-12-05 at 01 02 58
Screenshot 2024-12-05 at 01 03 46
Screenshot 2024-12-05 at 01 04 41
Screenshot 2024-12-05 at 01 05 36

**Comment by mojafa** [comment]
Screenshot 2024-12-05 at 01 20 51


you could use this merged PR as inspiration: https://github.com/daytonaio/content/pull/85/files


I'd also suggest, you let users create a new repo and then guide them on it step by step.

**Comment by mojafa** [comment]
Good stuff so far, you just picked a short cut, it's a guide, take your time and be very technical as it's a technical guide.

**Comment by Kiran1689** [comment]
Hey @mojafa 
Always check the latest code/files before review.

You are reviewing the old files..

Check here always for the latest committed 
https://github.com/daytonaio/content/pull/120/files

Most of them are already fixed..

**Comment by Kiran1689** [comment]
> you let users create a new repo and then guide them on it step by step.

I'll work on this

**Comment by mojafa** [comment]
Hey @Kiran1689 

I created a new workspace by using `daytona create https://github.com/daytonaio/content/pull/120` 
This will fetch your PR with it's latest changes. 


Interestingly I'm still seeing the same stuff.

See below:

Screenshot 2024-12-07 at 01 20 13

**Comment by Kiran1689** [comment]
@mojafa Check this 
- Added repository setup for step by step guide

**Comment by Kiran1689** [comment]
@mojafa can you please review this? So that I can start working on new issue..

**Comment by mojafa** [comment]
@Kiran1689 This is what I'm talking about now. 

The guide is comprehensive and covers all the points. Thank you.


And of course the code works...
Screenshot 2024-12-11 at 19 54 25

**Comment by Kiran1689** [comment]
@mojafa can you please provide your approval if everything is good..

**Comment by Kiran1689** [comment]
@nkkko Can you please check this?

**Comment by mojafa** [comment]
@Kiran1689 please be patient

**Pull request merged**